ISSUER_PROTOCOL.md

```diff
@@ -0,0 +1,345 @@
+  Wst = Sum(ej*Wsj) // over j from 0 to m
+  dleqProof = DLEQ2.P((X,T,S,Ws), (xs,ys)) // cs,us,vs
+  dleqorProof = DLEQOR2.P((X,T,S,W), (xb,yb)) // c0,c1,u0,u1,v0,v1
+  return dleqProof + dleqorProof
+```
+
+Input Serialization:
+
+The Trust Token Issuance Request contains an `IssueRequest` structure defined below.
+
+```
+struct {
+  uint16 count;
+  ECPoint nonces[count];
+} IssueRequest;
+```
+
+Output Serialization:
+
+The Trust Token Issuance Response contains an `IssueResponse` structure defined below.
+
+```
+struct {
+  opaque s<Nn>; // big-endian bytestring
+  opaque Wp<1..2^16-1>; // X9.62 Uncompressed Wp point.
+  opaque Wsp<1..2^16-1>; // X9.62 Uncompressed Wsp point.
+} SignedNonce;
+
+struct {
+  Scalar cs;
+  Scalar us;
+  Scalar vs;
+  Scalar c0;
+  Scalar c1;
+  Scalar u0;
+  Scalar u1;
+  Scalar v0;
+  Scalar v1;
+} DLEQProof;
+
+struct {
+  uint16 issued;
+  uint32 key_id = keyID;
+  SignedNonce signed[issued];
+  opaque proof<1..2^16-1>; // Length-prefixed form of DLEQProof.
+} IssueResponse;
+```
+
+### Redeem Function
+
+The _Redeem_ function corresponds to the **AT.VerValid** and **AT.ReadBit** stage of the PMBTokens protocol.
+
+Inputs:
+
+*   token (The trust token to redeem)
+*   client\_data (the client data sent as part of the redemption request to include in the SRR)
+*   redemptionTime (the redemption time from the client)
+*   secretKey (the secret key that should be used to sign this request, determined by the public metadata)
+*   lifetime (lifetime for the SRR)
+*   keys (dictionary from known key IDs to secret/public keys)
+*   srrKey (SRRSigningKey)
+
+Outputs:
+
+*   srr (the Signed Redemption Record, a binding of the client data and metadata to this redemption)
+*   signature (the signature over the SRR)
+
+```
+Redeem:
+  secretKey, publicKey = keys[token.key_id]
+  T = Ht(token.nonce)
+  if token.Ws != secretKey.xs*T+secretKey.ys*token.S:
+    return 0 // Invalid validity verification.
+  privateMetadata = 0
+  if token.W == secretKey.x0*T+secretKey.y0*token.S:
+    privateMetadata = 0
+  elseif token.W == secretKey.x1*T+secretKey.y1*token.S:
+    privateMetadata = 1
+  else:
+    return 0 // Internal Error
+  tokenHash = SHA256("TrustTokenV0 TokenHash"||token)
+  // Issuer-specific logic for encoding the private metadata bit.
+  encodedPrivate = EncodePrivateMetadata(privateMetadata)
+  srr = ConstructCBOR({
+    "metadata": { "public": token.key_id, "private": encodedPrivate},
+    "token-hash": tokenHash,
+    "client-data": client_data,
+    "expiry-timestamp": redemptionTime + lifetime
+  }) // Function to correctly encode a CBOR structure into a bytestring.
+  signature = Ed25519Sign(srrKey, srr)
+  return (srr, signature)
+```
+
+Input Serialization:
+
+The Trust Token Redemption Request contains a `RedemptionRequest` structure as defined below.
+
+```
+struct {
+  uint32 key_id;
+  opaque nonce<nonce_size>;
+  ECPoint S;
+  ECPoint W;
+  ECPoint Ws;
+} Token;
+
+struct {
+  opaque token<1..2^16-1>;
```

Comment:
Should we say in prose that `token` contains a `Token` structure?

Reply:
Done.